File: visa-payment-java/src/main/java/com/visa/paymentservice/model/Transaction.java
package com.visa.paymentservice.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "transactions")
public class Transaction {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "transaction_id")
    private Long transactionId;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "account_id", nullable = false)
    @NotNull(message = "Account is required")
    private Account account;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "operation_type_id", nullable = false)
    @NotNull(message = "Operation type is required")
    private OperationType operationType;
    
    @NotNull(message = "Amount is required")
    @Column(name = "amount", nullable = false, precision = 10, scale = 2)
    private BigDecimal amount;
    
    @Column(name = "event_date", nullable = false)
    private LocalDateTime eventDate;
    
    public Transaction() {
        this.eventDate = LocalDateTime.now();
    }
    
    public Transaction(Account account, OperationType operationType, BigDecimal amount) {
        this();
        this.account = account;
        this.operationType = operationType;
        this.amount = amount;
    }
    
    public Long getTransactionId() {
        return transactionId;
    }
    
    public void setTransactionId(Long transactionId) {
        this.transactionId = transactionId;
    }
    
    public Account getAccount() {
        return account;
    }
    
    public void setAccount(Account account) {
        this.account = account;
    }
    
    public OperationType getOperationType() {
        return operationType;
    }
    
    public void setOperationType(OperationType operationType) {
        this.operationType = operationType;
    }
    
    public BigDecimal getAmount() {
        return amount;
    }
    
    public void setAmount(BigDecimal amount) {
        this.amount = amount;
    }
    
    public LocalDateTime getEventDate() {
        return eventDate;
    }
    
    public void setEventDate(LocalDateTime eventDate) {
        this.eventDate = eventDate;
    }
}

File: visa-payment-go/internal/models/transaction.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type Transaction struct {
	ID              uint           `json:"transaction_id" gorm:"primaryKey" example:"1"`
	AccountID       uint           `json:"account_id" gorm:"not null" validate:"required" example:"1"`
	OperationTypeID uint           `json:"operation_type_id" gorm:"not null" validate:"required" example:"1"`
	Amount          float64        `json:"amount" gorm:"not null" validate:"required,gt=0" example:"50.00"`
	EventDate       time.Time      `json:"event_date" gorm:"not null"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `json:"-" gorm:"index"`

	// Relations
	Account       Account       `json:"account" gorm:"foreignKey:AccountID"`
	OperationType OperationType `json:"operation_type" gorm:"foreignKey:OperationTypeID"`
}

func (Transaction) TableName() string {
	return "transactions"
}
